### PR TITLE
Add health check

### PR DESCRIPTION
**Pull request by KungA**
Implementation of https://github.com/vostok/hosting/issues/70

### CODE REVIEW

**Comment by KungA** [comment]
Unfortunatelly, `ErrorCallback` is not enough for tracking ConfigurationProvider health: there are may be multiple TSettings types with different errors.

File: Vostok.Configuration/ConfigurationProvider.cs [hunk 1]
```diff
@@ -218,6 +219,10 @@ public bool TrySetupSourceFor([NotNull] Type settingsType, [NotNull] IConfigurat
         /// </summary>
         public bool HasSourceFor([NotNull] Type settingsType) => typeSources.ContainsKey(settingsType);
 
+        [CanBeNull]
+        public string GetHealthStatus() =>
```
Comment (KungA) [inline Vostok.Configuration/ConfigurationProvider.cs:223]: We can return `IEnumerable<(TSettings, HealthCheckResult)>`, but I don't see any reason for this complication.

File: Vostok.Configuration/ConfigurationProvider.cs [hunk 2]
```diff
@@ -55,7 +56,7 @@ public ConfigurationProvider(ConfigurationProviderSettings settings)
                 settings.SettingsCallback,
                 new ObservableBinder(new CachingBinder(new ValidatingBinder(settings.Binder ?? new DefaultSettingsBinder()))),
                 new SourceDataCache(settings.MaxSourceCacheSize),
-                new RetryingCurrentValueProviderFactory(settings.ValueRetryCooldown),
+                new RetryingCurrentValueProviderFactory(settings.ValueRetryCooldown, settings.ErrorCallback),
```
Comment (HolyPrapor) [inline Vostok.Configuration/ConfigurationProvider.cs:59]: Shouldn't we decorate this callback using `ErrorCallbackDecorator` for error deduplication?

File: Vostok.Configuration/ConfigurationProvider.cs (hunk 1 above)
Comment (HolyPrapor) [reply Vostok.Configuration/ConfigurationProvider.cs:223]: You are correct, it's not very useful.

However, I have a couple of concerns here:
1. An implicit conversion of (string == null <=> there is an error). 
2. We are unable to add new information to the result of this method in the future. (Or at least we will not be able to provide a convenient interface for this).

I think it's reasonable enough to create a container `ConfigurationHealthCheck`, which can solve both of these problems.
 
But it's up to you on this.

Comment (KungA) [reply Vostok.Configuration/ConfigurationProvider.cs:223]: Yeah, I even have new case: health check may be Failing on first settings update and Degraded on next updates (after a successful one).

File: Vostok.Configuration/ConfigurationProvider.cs (hunk 2 above)
Comment (KungA) [reply Vostok.Configuration/ConfigurationProvider.cs:59]: I've just moved this from parameter to constructor :)

But probably not: you want to see error every time it happens.

Comment (HolyPrapor) [reply Vostok.Configuration/ConfigurationProvider.cs:59]: I meant that the parameter option passed not the original `settings.ErrorCallback`, but `ErrorCallbackDecorator` one. 
This conversion is done in the constructor of ConfigurationProvider.

But now you pass the original `settings.ErrorCallback`.


Also, that's not 'deduplication' in a purest form, it's just 'error throttling', so I think this is still quite useful.

File: Vostok.Configuration/Helpers/ConfigurationHealthCheckResult.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
+using JetBrains.Annotations;
+
+namespace Vostok.Configuration.Helpers;
+
```
Comment (HolyPrapor) [inline Vostok.Configuration/Helpers/ConfigurationHealthCheckResult.cs:4]:
```suggestion
[PublicAPI]
```

**Review by HolyPrapor — APPROVED** [review]

**Pull request merged**